core/embed/lib/image.c

```diff
@@ -224,35 +243,106 @@ secbool check_image_contents(const image_header *const hdr, uint32_t firstskip,
 
-  int chunk = 1;
-
-  while (remaining > 0) {
-    data = flash_area_get_address(area, chunk * IMAGE_CHUNK_SIZE,
-                                  IMAGE_CHUNK_SIZE);
+    const void *data = flash_area_get_address(area, offset, bytes_to_check);
     if (!data) {
       return secfalse;
     }
-    if (sectrue != check_single_hash(hdr->hashes + chunk * 32, data,
-                                     MIN(remaining, IMAGE_CHUNK_SIZE))) {
+
+    size_t hash_offset = (offset / IMAGE_CHUNK_SIZE) * 32;
+    if (sectrue !=
+        check_single_hash(hdr->hashes + hash_offset, data, bytes_to_check)) {
+      return secfalse;
+    }
+
+    offset += bytes_to_check;
+  }
+
+  // Check the padding to the end of the area
+  end_offset = flash_area_get_size(area);
+
+  if (offset < end_offset) {
+    // Use the first byte in the checked area as the expected padding byte
+    // Firmware is always padded with 0xFF, while the bootloader might be
+    // padded with 0x00 as well
+    uint8_t expected_byte = *(
+        (const uint8_t *)flash_area_get_address(area, offset, sizeof(uint8_t)));
+
+    if (expected_byte != 0x00 && expected_byte != 0xFF) {
       return secfalse;
     }
-    chunk++;
-    remaining -= IMAGE_CHUNK_SIZE;
+
+    while (offset < end_offset) {
+      size_t bytes_to_check = MIN(
+          IMAGE_CHUNK_SIZE - (offset % IMAGE_CHUNK_SIZE), end_offset - offset);
+
+      const uint8_t *data =
+          (const uint8_t *)flash_area_get_address(area, offset, bytes_to_check);
+      if (!data) {
+        return secfalse;
+      }
+
+      for (size_t i = 0; i < bytes_to_check; i++) {
+        if (data[i] != expected_byte) {
+          return secfalse;
+        }
+      }
+
+      offset += bytes_to_check;
+    }
+  }
+
+  return sectrue;
+}
+
+secbool check_firmware_header(const uint8_t *header, size_t header_size,
+                              firmware_header_info_t *info) {
+  // parse and check vendor header
+  vendor_header vhdr;
+  if (sectrue != read_vendor_header(header, &vhdr)) {
+    return secfalse;
+  }
+  if (sectrue != check_vendor_header_keys(&vhdr)) {
+    return secfalse;
+  }
+
+  // parse and check image header
+  const image_header *ihdr;
+  if ((ihdr = read_image_header(header + vhdr.hdrlen, FIRMWARE_IMAGE_MAGIC,
+                                FIRMWARE_IMAGE_MAXSIZE)) == NULL) {
+    return secfalse;
+  }
+  if (sectrue !=
+      check_image_header_sig(ihdr, vhdr.vsig_m, vhdr.vsig_n, vhdr.vpub)) {
+    return secfalse;
+  }
+
+  // copy vendor string
+  info->vstr_len = MIN(sizeof(info->vstr), vhdr.vstr_len);
+  if (info->vstr_len > 0) {
+    memcpy(info->vstr, vhdr.vstr, info->vstr_len);
   }
 
+  // copy firmware version
+  info->ver_major = ihdr->version & 0xFF;
+  info->ver_minor = (ihdr->version >> 8) & 0xFF;
+  info->ver_patch = (ihdr->version >> 16) & 0xFF;
+
+  // copy fingerprint from image header
+  memcpy(&info->fingerprint, ihdr->sig, sizeof(info->fingerprint));
```

Comment:
This copies the signature instead. I think you need to call `get_image_fingerprint` to obtain the fingerprint.

Reply:
The input to this function consists of just the vendor and image header received in the RebootToBootloader message. We do not have the complete firmware binary at this point.

Reply:
probably better to leave the field empty or even delete the field, since we can't know the fingerprint from headers only

Reply:
of course we can, the fingerprint is a hash of the header excluding the signature fields. the argument to `get_image_fingerprint` is just `hdr`

Reply:
fixed, now passing proper fingerprint, see 8a88f2c8bab88a161635fe1bdac5f9bdd0a099c0